### PR TITLE
Add new constraints to ThresholdOptimizer and add InterpolatedThresholder

### PR DESCRIPTION
**Pull request by MiroDudik**

### CODE REVIEW

File: fairlearn/postprocessing/_constants.py [hunk 1]
```diff
@@ -11,6 +13,45 @@
 
 DEMOGRAPHIC_PARITY = "demographic_parity"
 EQUALIZED_ODDS = "equalized_odds"
+ACCURACY_SCORE = "accuracy_score"
+
+def confusion_matrix_summary(*, true_positives, false_positives, true_negatives, false_negatives):
```
Comment (romanlutz) [inline fairlearn/postprocessing/_constants.py:18]: That's not really a "constant" :-)

File: test/unit/postprocessing/test_threshold_optimization.py [hunk 2]
```diff
@@ -150,24 +150,19 @@ def test_threshold_optimization_different_input_lengths(data_X_y_sf, constraints
...
                                                    sensitive_features_transform):
     y = y_transform(labels_ex)
     sensitive_features = sensitive_features_transform(sensitive_features_ex1)
-    scores = score_transform(scores_ex)
-    estimator = ThresholdOptimizer(estimator=ExamplePredictor(scores),
+    estimator = ThresholdOptimizer(estimator=PassThroughPredictor(),
                                    constraints=DEMOGRAPHIC_PARITY)
-    estimator.fit(X_ex, y, sensitive_features=sensitive_features)
-
+    estimator.fit(pd.DataFrame(scores_ex), y, sensitive_features=sensitive_features)
```
Comment (romanlutz) [inline test/unit/postprocessing/test_threshold_optimization.py:161]: Why was this required? Did we break something about supported data types?

File: test/unit/postprocessing/test_threshold_optimization.py [hunk 3]
```diff
@@ -150,24 +150,19 @@ def test_threshold_optimization_different_input_lengths(data_X_y_sf, constraints
...
 
-@pytest.mark.parametrize("score_transform", candidate_Y_transforms)
 @pytest.mark.parametrize("y_transform", candidate_Y_transforms)
 @pytest.mark.parametrize("sensitive_features_transform", candidate_A_transforms)
-def test_threshold_optimization_demographic_parity(score_transform, y_transform,
+def test_threshold_optimization_demographic_parity(y_transform,
                                                    sensitive_features_transform):
     y = y_transform(labels_ex)
     sensitive_features = sensitive_features_transform(sensitive_features_ex1)
-    scores = score_transform(scores_ex)
```
Comment (romanlutz) [inline test/unit/postprocessing/test_threshold_optimization.py:160]: I'm fine with the other changes, but score_transform should still be used, otherwise we don't have test coverage for other cases.

File: test/unit/postprocessing/test_curve_utilities.py [hunk 4]
```diff
@@ -89,24 +89,37 @@ def test_calculate_roc_points():
...
 
-    roc_points = _calculate_roc_points(grouped_data, "A")
+    roc_points = _calculate_roc_points(grouped_data, "A", flip=True)
     expected_roc_points = pd.DataFrame({
-        "x": [0, 0.25, 0.5, 0.5, 1],
-        "y": [0, 1/3,  2/3, 1,   1],
+        "x": [0, 1, 0.75, 0.25, 0.5, 0.5, 0.5, 0.5, 1, 0],
+        "y": [0, 1, 2/3,  1/3,  1/3, 2/3, 0,   1,   1, 0],
         "operation": [ThresholdOperation('>', np.inf),
+                      ThresholdOperation('<', np.inf),
```
Comment (romanlutz) [inline test/unit/postprocessing/test_curve_utilities.py:97]: Why change the original example?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 5]
```diff
@@ -93,9 +214,16 @@ def fit(self, X, y, *, sensitive_features, **kwargs):
         if self.estimator is None:
             raise ValueError(ESTIMATOR_ERROR_MESSAGE)
 
-        if self.constraints not in _SUPPORTED_CONSTRAINTS:
+        if self.constraints not in SUPPORTED_CONSTRAINTS:
             raise ValueError(NOT_SUPPORTED_CONSTRAINTS_ERROR_MESSAGE)
 
+        if self.constraints == EQUALIZED_ODDS:
+            if self.objective not in SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS:
+                raise ValueError(NOT_SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS_ERROR_MESSAGE)
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:222]: There should be a test case for this (and any error message really)

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (romanlutz) [inline fairlearn/postprocessing/_constants.py:18]: Any chance we could reuse something that already exists for this? Perhaps https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html is of use?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 6]
```diff
@@ -38,12 +39,131 @@
...
 ESTIMATOR_ERROR_MESSAGE = "'estimator' cannot be `None`."
 
-_SUPPORTED_CONSTRAINTS = [DEMOGRAPHIC_PARITY, EQUALIZED_ODDS]
-
 logger = logging.getLogger(__name__)
 
+# All allowed utilities must have values between 0 and 1 and attain both extremes as
+# the threshold goes from -Inf to Inf.
+SUPPORTED_UTILITIES = {
+    'selection_rate',
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:47]: This is equivalent to DP. Why add it?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 7]
```diff
@@ -38,12 +39,131 @@
...
+    "Currently only the following constraints are supported: {}.".format(
+        ", ".join(SUPPORTED_CONSTRAINTS)))
+NOT_SUPPORTED_OBJECTIVES_ERROR_MESSAGE = (
+    "Currently only the following objectives are supported: {}.".format(
+    ", ".join(METRIC_DICT.keys())))
+NOT_SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS_ERROR_MESSAGE = (
+    "Currently only the following objectives are supported: {}.".format(
+    ", ".join(SUPPORTED_OBJECTIVES_FOR_EQUALIZED_ODDS)))
+
+class InterpolatedThresholder(BaseEstimator, MetaEstimatorMixin):
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:72]: I'm not sure I like that it's in the same file. Makes the entire file more crowded with things. :-(
So far, if I search for `fit` in this file I know where I end up. Now I have to scroll back up and search for the class name.

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 8]
```diff
@@ -38,12 +39,131 @@
...
+    """Binary predictor that thresholds continuous predictions of a base estimator.
+
+    At prediction time, the predictor takes as input both standard and sensitive features.
+    Based on the values of sensitive features, it then applies a randomized thresholding
+    transformation according to the provided `interpolation_dict`.
+    
+    :param estimator: base estimator
+    :param dict interpolation_dict: maps sensitive feature values to `Bunch` that describes the
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:80]: Is this a new way of writing the doc, or did you mean
```
:param interpolation_dict: …
:type interpolation_dict: dict
```
?

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py:18]: shall we rename `_constants` -> `_base`?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 9]
```diff
@@ -38,12 +39,131 @@
...
+        :rtype: numpy.ndarray
+        """
+        check_is_fitted(self)
+        _, _, sensitive_feature_vector = _validate_and_reformat_input(
+            X, y=None, sensitive_features=sensitive_features, expect_y=False,
+            enforce_binary_labels=True)
+        base_predictions = np.array(self.estimator_.predict(X))
+        
+        positive_probs = 0.0*base_predictions
+        for a, interpolation in self.interpolation_dict.items():
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:133]: Can we use `sensitive_feature_value` or at least `sf` instead of `a`?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 10]
```diff
@@ -38,12 +39,131 @@
...
+        
+    def predict(self, X, *, sensitive_features, random_state=None):
+        """Provide a prediction for the given input data.
+
+        Note that this is non-deterministic, due to the nature of the
+        interpolation.
 
-class ThresholdOptimizer(BaseEstimator, ClassifierMixin, MetaEstimatorMixin):
+        :param X: Feature data
+        :type X: numpy.ndarray or pandas.DataFrame
```
Comment (romanlutz) [inline fairlearn/postprocessing/_threshold_optimizer.py:152]: Can we also add the other params in the documentation?

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_curve_utilities.py:97]: Keeping the existing tests means we didn't regress anywhere so that would be very important for a PR of this size. You can add as many additional examples as you can write :-)

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py:18]: This is doing something very different. It is taking an already calculated confusion matrix and filling in the names of implied fields. Would it help to rename this utility function?

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_curve_utilities.py:97]: The logic of `_calculate_roc_points` has changed--that's why the change here. New logic is more generic, so I think I prefer it.

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 3 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:160]: I don't think we need to vary score transform in this instance. The base estimator is supposed to be implementing an sklearn estimator and those are always supposed to return 1D-arrays (not lists and not DataFrames). I don't think we need to try to be more general than sklearn here.

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 2 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:161]: This is because I'm now working with `PassThroughPredictor` instead of `ExamplePredictor`.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 7 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:72]: I don't mind pulling out into a separate file.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 8 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:80]: this is supported by sphinx and looks more natural to me... i'm even thinking we should consider going numpy style like sklear, because it reads nicer when you call `help` in interactive python.

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_constants.py:18]: Base kind of makes me think "base class". So far everything like that is named utilities. The constants just don't belong in the same file, though. Can you put the functions/logic in a utilities (or other name) file?

Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py:18]: i see your point. let me see what i can do. i resist creating a lot of small files...

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_curve_utilities.py:97]: I'm actually not entirely sure whether I understand it anymore. These used to be ordered in an understandable way, but now it's hard to grasp how this all came to be.

You mean the old example doesn't work anymore? Because it's hard to tell if we break something if we change the test logic itself.

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 3 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:160]: On a second thought, let me try to add this back. I don't trust people to implement sklearn estimators 100% faithfully to the spec, and it would be nice to cover these simple cases.

Comment (romanlutz) [reply test/unit/postprocessing/test_threshold_optimization.py:160]: Then we need to adjust the expected data types to exclude those. At every other point we've parametrized over the different data types.

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 2 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_threshold_optimization.py:161]: This goes back to the same point as in the other comment. If we allow more data types we should test the entire matrix of data types. If we don't allow it we need to adjust the places where we've said we do.

Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:161]: This is a different issue. The feature matrix should always be a pass-through, we don't touch it, we just pass it on to the base estimator. We cannot guarantee that the base estimator will work well with one or another kind of feature encoding (this is what burnt us with LightGBM). Maybe best is to remove all of the validation that we do around X. But that should be a separate PR.

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_curve_utilities.py:97]: [exactly... let me re-order them in an understandable way again]

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 8 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:80]: @adrinjalali recently pointed that out in an issue, I think #314

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 6 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:47]: Or rather: why add this one and not others like equality of opportunity and other duplicate names?

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_constants.py:18]: I kind of understood why it's there better after reading the entire PR. You're mapping the various objectives to these metrics inside postprocessing, so this is required. I think it's fine the way it is, but it would be awesome to have a short description saying why this is needed (i.e. the second sentence of this very comment, but perhaps more nicely written?)

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_curve_utilities.py:97]: OK, thanks for elaborating!
Would it be possible to keep the old case (probably with adjusted result) and add this new case additionally?

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 2 above)
Comment (romanlutz) [reply test/unit/postprocessing/test_threshold_optimization.py:161]: Wow I completely confused myself... this is passed as the feature matrix, of course, not as scores. 

What burnt us with LightGBM was that we changed X (and that was in reductions, not postprocessing). That's not the case here, though. You should be able to pass in any kind of format that we allow for X, and therefore we should also be testing any kind of format as input.

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_constants.py:18]: If you find a more suitable name than constants or base I don't mind that either... :-)

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 6 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:47]: I'd like to limit the duplication unless there is a strong reason against it. The "systematic" terminology is "selection_rate", but "demographic_parity" is extremely common, so it's added as a synonym (and for the backward compatibility).

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_curve_utilities.py:97]: Technically, the behavior of `_calculate_roc_points` is now different (it produces lists twice as long as before), so this is a breaking change. If I left the previous test unmodified it would fail. What I have here is an equivalent test (on the exactly same input data). Do you want to separate out the old functionality of the previous  `_calculate_roc_points`? (I'm not sure I see why.)

Comment (romanlutz) [reply test/unit/postprocessing/test_curve_utilities.py:97]: I feel like we already agree, but I'm expressing myself badly :-)
What I meant to say is that when you change logic like in `_calculate_roc_points`, then it's natural that the test case expectations may need to be adjusted. For example, the expected result may be twice as long as you mention. However, in this particular case the input data itself was also changed (or at least it looks like that to me), which means it's not actually possible for us to understand how it affects the test case without calculating everything by hand. From the diff I expected to see the identical input data and then an expected result list that's twice as long.
Perhaps git/GitHub is playing tricks on me and you didn't change the test input data?

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 6 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:47]: Sure, if you feel strongly then go with that. My question was rather out of curiosity.

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (adrinjalali) [reply fairlearn/postprocessing/_constants.py:18]: `extended_confusion_matrix`?

File: fairlearn/postprocessing/_roc_curve_utilities.py [hunk 11]
```diff
@@ -4,13 +4,17 @@
...
+    LABEL_KEY, SCORE_KEY, P0_KEY, P1_KEY,
+    METRIC_DICT,
+    confusion_matrix_summary)
 from ._threshold_operation import ThresholdOperation
 
 DEGENERATE_LABELS_ERROR_MESSAGE = "Degenerate labels for sensitive feature value {}"
 
 
-def _get_roc(data, sensitive_feature_value, flip=True):
+def _get_roc(data, sensitive_feature_value, flip=False,
```
Comment (adrinjalali) [inline fairlearn/postprocessing/_roc_curve_utilities.py:16]: is this very different from https://scikit-learn.org/dev/modules/generated/sklearn.metrics.roc_curve.html#sklearn.metrics.roc_curve?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 12]
```diff
@@ -38,12 +39,131 @@
...
+        If `prefit` is set to `True` then the base estimator is kept as is.
+        Otherwise it is fitted from the provided arguments.
+        """
+        if self.estimator is None:
+            raise ValueError(ESTIMATOR_ERROR_MESSAGE)
+        
+        if not self.prefit:
+            self.estimator_ = clone(self.estimator).fit(X, y, **kwargs)
+        else:
+            self.estimator_ = self.estimator
```
Comment (adrinjalali) [inline fairlearn/postprocessing/_threshold_optimizer.py:110]: this should also check if the estimator is already fitted. Otherwise it should refit anyway.

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 13]
```diff
@@ -38,12 +39,131 @@
...
+            of the two numbers in each tuple needs to add up to 1.
+        :rtype: numpy.ndarray
+        """
+        check_is_fitted(self)
+        _, _, sensitive_feature_vector = _validate_and_reformat_input(
+            X, y=None, sensitive_features=sensitive_features, expect_y=False,
+            enforce_binary_labels=True)
+        base_predictions = np.array(self.estimator_.predict(X))
+        
+        positive_probs = 0.0*base_predictions
```
Comment (adrinjalali) [inline fairlearn/postprocessing/_threshold_optimizer.py:132]: maybe [`np.zeros_like`](https://numpy.org/doc/stable/reference/generated/numpy.zeros_like.html?highlight=zeros_like#numpy.zeros_like)?

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 14]
```diff
@@ -65,12 +185,13 @@ class ThresholdOptimizer(BaseEstimator, ClassifierMixin, MetaEstimatorMixin):
     """
 
     def __init__(self, *, estimator=None,
-                 constraints=DEMOGRAPHIC_PARITY, grid_size=1000, flip=True,
+                 constraints=DEMOGRAPHIC_PARITY, objective=ACCURACY_SCORE, grid_size=1000, flip=True,
```
Comment (adrinjalali) [inline fairlearn/postprocessing/_threshold_optimizer.py:188]: the issue with having these constants such as `ACCURACY_SCORE` is that when user sees the autocomplete in their IDE, it would look less confusing if they saw `"accuracy_score"` I think.

File: fairlearn/postprocessing/_threshold_optimizer.py [hunk 15]
```diff
@@ -170,16 +282,11 @@ def _pmf_predict(self, X, *, sensitive_features):
...
-            self._post_processed_predictor_by_sensitive_feature, sensitive_feature_vector,
-            self.estimator_.predict(X))
-        return np.array([[1.0 - p, p] for p in positive_probs])
-
-    def _threshold_optimization_demographic_parity(self, sensitive_features, labels, scores,
-                                                   grid_size=1000, flip=True):
+        return self.interpolated_thresholder_._pmf_predict(
+            X, sensitive_features=sensitive_features)
+        
+    def _threshold_optimization_for_utility(
```
Comment (adrinjalali) [inline fairlearn/postprocessing/_threshold_optimizer.py:288]: a better name here maybe

File: test/unit/postprocessing/conftest.py [hunk 16]
```diff
@@ -113,6 +113,14 @@ def predict(self, X):
...
 
 
+class PassThroughPredictor(BaseEstimator, ClassifierMixin):
+    def fit(self, X, y=None, **kwargs):
+        return self
+
+    def predict(self, X):
+        return X[0]
+
+
```
Comment (adrinjalali) [inline test/unit/postprocessing/conftest.py:123]: I think all these would much better fit in the test file themselves instead of in the `conftest.py`. I find it odd to have these here.

File: fairlearn/postprocessing/_constants.py [hunk 17]
```diff
@@ -11,6 +13,45 @@
 
 DEMOGRAPHIC_PARITY = "demographic_parity"
 EQUALIZED_ODDS = "equalized_odds"
+ACCURACY_SCORE = "accuracy_score"
```
[the remark this answers is not in the document]
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py:16]: I think that our names of metrics are now stable enough that we don't need these three constants as long as we have tests that exercise all valid strings.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: This actually points out to a deeper issue.

I tried to do it the way you suggested first, but running with `LGBMClassifier` was triggering refitting even though it was already fitted. Obviously they are not implementing sklearn spec correctly, but that might end up being quite common when people create simple wrappers around their non-sklearn classifiers to pull into our library. I am considering two options here:
- Don't make `InterpolatedThresholder` a sklearn estimator; we would not bring in the mixins and would only implement `predict`, but not `fit`. (Alas, I would use the pretty printing of `BaseEstimator.__repr__`
- Throw a warning or even an exception when there's a mismatch between `prefit` and `check_fitted` (and required a manual override for an exception, e.g., by setting an additional argument? or is that too much?)
Thoughts?

File: test/unit/postprocessing/conftest.py (hunk 16 above)
Comment (romanlutz) [reply test/unit/postprocessing/conftest.py:123]: Perhaps this continues an odd pattern we've accidentally established...
I agree with @adrinjalali, though, and will make a mental note to clean other similar cases up.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 15 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:288]: I love this comment, because the name makes perfect sense to me, but it shows that it's not quite as clear as I thought! 
It does perform threshold optimization for a provided utility, though, so I wonder how we could find a better name.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 14 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:188]: Fair point! Does sklearn repeat all the strings all over the place, or is there a better pattern that we just haven't come across?

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: I dislike not conforming to sklearn because other packages don't follow conventions properly.
I also dislike adding yet another argument, because that would just completely confuse me as the user.

Let me see whether I can write up expected behavior (IMO)
- `prefit=True` and `check_fitted` returns True: no additional fitting
- `prefit=True` and `check_fitted` returns False: user tells us explicitly that it's fitted, so don't fit again, but log warning and if it fails on `predict` provide an explanation. This seems like a typical scenario of "user error" to me. If a user provides `prefit=True` we should never call `fit`.
- `prefit=False` and `check_fitted` returns True: call `fit` anyway because the user seems to want us to refit it.
- `prefit=False` and `check_fitted` returns False: call `fit`

What do you think?

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/postprocessing/_constants.py:18]: I rather like the verbosity here (and in general), but I'm not too invested in this.

File: fairlearn/postprocessing/_constants.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (romanlutz) [reply fairlearn/postprocessing/_constants.py:16]: and perhaps we should put them into the doc for ThresholdOptimizer to clearly say what the options are

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (adrinjalali) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: The issue with `prefit=True` and warning or raising is that `clone` will keep `prefit=True`, but the underlying estimator will be a fresh unfitted one. To me it's a straightforward issue, `LGBMClassifier` has an issue which can very easily be fixed, so I'd open an issue on https://github.com/microsoft/lightgbm and most probably have it fixed. It'll probably be better received if the issue report comes from withing Microsoft (gut feeling, not sure about your politics ;) )

File: fairlearn/postprocessing/_constants.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (adrinjalali) [reply fairlearn/postprocessing/_constants.py:16]: ideally the docstring should include all the options, and a link to a "user guide" is beneficial for further explanation ;)

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 14 above)
Comment (adrinjalali) [reply fairlearn/postprocessing/_threshold_optimizer.py:188]: I guess it's just a norm that different communities follow. And in a dynamically typed lang like python, having the strings makes sense IMO. Alternatively, you could go for an `Enum`, but I haven't seen that often in python, and it can complicate things. What makes sense to me, is ['Literal`](https://www.python.org/dev/peps/pep-0586/) which we probably will see more of once people start adopting Python 3.8 standards.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (romanlutz) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: That helps with LightGBM, but that won't fix the same issue for other packages.

This is what I filed with them: https://github.com/microsoft/LightGBM/issues/3014

Comment (adrinjalali) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: Well, if you go with a nice sklearn compat, then it makes sense for the user to work in that ecosystem I'd say.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 15 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:288]: I will be renaming `utility` to `simple_metrics` as opposed to `equalized_odds` (which is the one "not simple" metric that requires a separate treatment).

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: I kinda want to have a cake and eat it too. @romanlutz : your proposal exactly matches mine.

A slight elaboration:
- if `prefit=True` and `check_fitted=True`: no additional fitting
- if `prefit=True` and `check_fitted=True`: no additional fitting, but give a warning--the warning should say that if the user is running crossvalidation etc., they should do so with `prefit=False` and if the estimator is already fitted, they should file the issue to make it sklearn compliant. 
- if `prefit=False`: always call `fit`

@adrinjalali :  I'm also a bit worried about a silent failure (where the user is setting `prefit=True`, but is unaware that the base estimator is being refitted). This proposal addresses that issue too.

Comment (adrinjalali) [reply fairlearn/postprocessing/_threshold_optimizer.py:110]: I was checking `SelectFromModel` in sklearn, and apparently there we even raise if the user calls `fit` and `prefit=True`, which means users can't give an estimator with `prefit=True` to a `GridSearchCV` for instance. So a warning, or even raising, would be consistent with that.

File: fairlearn/_input_validation.py [hunk 18]
```diff
@@ -54,9 +54,9 @@ def _validate_and_reformat_input(X, y=None, expect_y=True, enforce_binary_labels
     if y is not None:
         # calling check_X_y with a 2-dimensional y causes a warning, so ensure it is 1-dimensional
         if isinstance(y, np.ndarray) and len(y.shape) == 2 and y.shape[1] == 1:
-            y = y.squeeze()
+            y = y.reshape(-1)
```
Comment (romanlutz) [inline fairlearn/_input_validation.py:57]: why not squeeze anymore?

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py]: I'm moving it elsewhere and renaming to `_extend_confusion_matrix`.

File: fairlearn/_input_validation.py (hunk 18 above)
Comment (MiroDudik) [reply fairlearn/_input_validation.py:57]: it reduced the number of dimensions too much, down to a scalar, instead down to a one-dimensional array.

File: fairlearn/postprocessing/_roc_curve_utilities.py (hunk 11 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_roc_curve_utilities.py:54]: This function is in some ways more general than the sklearn's (but in some ways less general). We allow picking  `x_metric` and `y_metric` that are different from FPR and TPR, and also we allow the ability to flip the decision if that improves `y_metric`. I think we should rename this from `roc_curve` to a `tradeoff_curve`. I'll update.

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 3 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:160]: `score_ex` is now again included and I've also added a comment about the `PassThroughPredictor`

File: test/unit/postprocessing/test_threshold_optimization.py (hunk 2 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_threshold_optimization.py:177]: The previous version was only trying different transformed versions of `y`, `scores_ex` and `sensitive_features`, so I've retained that functionality. Let me know if you think I should still be changing anything here.

File: test/unit/postprocessing/test_curve_utilities.py (hunk 4 above)
Comment (MiroDudik) [reply test/unit/postprocessing/test_curve_utilities.py]: Please take a look again. To make it clear how I changed things, the additional returned points are written in the form `.0` and `.5` rather than `0` and `0.5` etc.

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py]: Adding a more detailed docstring.

File: fairlearn/postprocessing/_constants.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py]: I'm implementing all the docstring suggestions, but not pointing to the User Guide yet--until we write that content in a future PR.

File: fairlearn/postprocessing/_constants.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiroDudik) [reply fairlearn/postprocessing/_constants.py]: i'll stick to the verbose form now, but making this method private, so we can revisit a more concise API if we want to make it public.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 12 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py]: Great, I'll then implement my last suggestion.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 14 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py]: Moving to explicit strings since metric names are now stable.

File: test/unit/postprocessing/conftest.py (hunk 16 above)
Comment (MiroDudik) [reply test/unit/postprocessing/conftest.py]: Done.

File: fairlearn/postprocessing/_threshold_optimizer.py (hunk 13 above)
Comment (MiroDudik) [reply fairlearn/postprocessing/_threshold_optimizer.py]: I ended up keeping as is, because `base_predictions` is `int`, but `positive_probs` needs to be `float`.

**Comment by romanlutz** [comment]
This PR looks great. Is this still WIP?

**Comment by MiroDudik** [comment]
Actually--you are right. This is definitely no longer WIP. The "only" missing piece I believe are tests for the new functionality.

File: fairlearn/postprocessing/_interpolated_thresholder.py [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+
+
+class InterpolatedThresholder(BaseEstimator, MetaEstimatorMixin):
+    """Binary predictor that thresholds continuous predictions of a base estimator.
+
+    At prediction time, the predictor takes as input both standard and sensitive features.
+    Based on the values of sensitive features, it then applies a randomized thresholding
+    transformation according to the provided `interpolation_dict`.
+
+    :param estimator: base estimator
```
Comment (romanlutz) [inline fairlearn/postprocessing/_interpolated_thresholder.py:26]: Did you want to use numpydoc all over this PR?

**Comment by MiroDudik** [comment]
@adrinjalali : This PR overlaps with #282 and #342 . It fixes some of the issues noted in #282 (such as `prefit`), but not all of them. Is it okay if I go ahead and merge this one? I can help you incorporate the additional changes from #282.

**Comment by adrinjalali** [comment]
Yes of course. My PR requires some changes in scikit-learn and that's why it's on hold. Please don't wait for that one. I'll probably need to open a new one once certain issues are fixed on the sklearn side.
